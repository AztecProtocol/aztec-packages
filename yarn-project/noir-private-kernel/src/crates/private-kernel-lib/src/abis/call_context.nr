use crate::abis::function_selector::FunctionSelector;
use crate::address::{EthAddress,Address};
use dep::aztec::constants_gen;

struct CallContext{
    msg_sender : Address,
    storage_contract_address : Address,
    portal_contract_address : EthAddress,

    function_selector : FunctionSelector,

    is_delegate_call : bool,
    is_static_call : bool,
    is_contract_deployment : bool,
}

impl CallContext {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.msg_sender.to_field(),
            self.storage_contract_address.to_field(),
            self.portal_contract_address.to_field(),
            self.function_selector.to_field(),
            self.is_delegate_call as Field,
            self.is_static_call as Field,
            self.is_contract_deployment as Field,
        ], constants_gen::GENERATOR_INDEX__CALL_CONTEXT)[0]
    }

    fn assert_is_zero(self) {
        assert(self.msg_sender.to_field() == 0);
        assert(self.storage_contract_address.to_field() == 0);
        assert(self.portal_contract_address.to_field() == 0);
        assert(self.function_selector.to_field() == 0);
        assert(self.is_delegate_call == false);
        assert(self.is_static_call == false);
        assert(self.is_contract_deployment == false);
    }
}
