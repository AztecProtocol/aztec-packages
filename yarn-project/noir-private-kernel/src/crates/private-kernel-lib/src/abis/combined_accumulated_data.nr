use crate::mocked::AggregationObject;
use crate::abis::new_contract_data::NewContractData;
use crate::abis::optionally_revealed_data::OptionallyRevealedData;
use crate::abis::public_data_update_request::PublicDataUpdateRequest;
use crate::abis::public_data_read::PublicDataRead;
use dep::aztec::constants_gen::{
    MAX_READ_REQUESTS_PER_TX,
    MAX_NEW_COMMITMENTS_PER_TX,
    MAX_NEW_NULLIFIERS_PER_TX,
    MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX,
    MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX,
    MAX_NEW_L2_TO_L1_MSGS_PER_TX,
    MAX_NEW_CONTRACTS_PER_TX,
    MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX,
    MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX,
    MAX_PUBLIC_DATA_READS_PER_TX
};
use crate::hash::NUM_FIELDS_PER_SHA256;
use crate::utils::bounded_vec::BoundedVec;

struct CombinedAccumulatedData {
    aggregation_object : AggregationObject,

    read_requests : BoundedVec<Field,MAX_READ_REQUESTS_PER_TX>,

    new_commitments : BoundedVec<Field, MAX_NEW_COMMITMENTS_PER_TX>,
    new_nullifiers : BoundedVec<Field, MAX_NEW_NULLIFIERS_PER_TX>,
    nullified_commitments : BoundedVec<Field, MAX_NEW_NULLIFIERS_PER_TX>,

    private_call_stack : BoundedVec<Field, MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX>,
    public_call_stack : BoundedVec<Field, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX>,
    new_l2_to_l1_msgs : BoundedVec<Field, MAX_NEW_L2_TO_L1_MSGS_PER_TX>,

    encrypted_logs_hash : [Field; NUM_FIELDS_PER_SHA256],
    unencrypted_logs_hash : [Field; NUM_FIELDS_PER_SHA256],

    // Here so that the gas cost of this request can be measured by circuits, without actually needing to feed in the
    // variable-length data.
    encrypted_log_preimages_length : Field,
    unencrypted_log_preimages_length : Field,

    new_contracts : BoundedVec<NewContractData, MAX_NEW_CONTRACTS_PER_TX>,

    optionally_revealed_data : BoundedVec<OptionallyRevealedData, MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX>,

    public_data_update_requests : BoundedVec<PublicDataUpdateRequest, MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX>,

    public_data_reads : BoundedVec<PublicDataRead, MAX_PUBLIC_DATA_READS_PER_TX>
}
