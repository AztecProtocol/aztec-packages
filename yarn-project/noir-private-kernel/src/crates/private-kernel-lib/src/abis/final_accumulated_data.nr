use crate::mocked::AggregationObject;
use crate::hash::NUM_FIELDS_PER_SHA256;
use dep::aztec::constants_gen::{MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX,MAX_NEW_CONTRACTS_PER_TX,MAX_NEW_L2_TO_L1_MSGS_PER_TX,MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX,MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX,MAX_NEW_NULLIFIERS_PER_TX,MAX_NEW_COMMITMENTS_PER_TX};
use crate::abis::new_contract_data::NewContractData;
use crate::abis::optionally_revealed_data::OptionallyRevealedData;
use crate::utils::bounded_vec::BoundedVec;

struct FinalAccumulatedDataBuilder {

    aggregation_object : AggregationObject,

    new_commitments : BoundedVec<Field, MAX_NEW_COMMITMENTS_PER_TX>,
    new_nullifiers : BoundedVec<Field, MAX_NEW_NULLIFIERS_PER_TX>,
    nullified_commitments : BoundedVec<Field, MAX_NEW_NULLIFIERS_PER_TX>,

    // For pending nullifiers, we have:
    // nullifiedCommitments[j] != 0 <==> newNullifiers[j] nullifies nullifiedCommitments[j]

    private_call_stack : BoundedVec<Field, MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX>,
    public_call_stack : BoundedVec<Field, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX>,
    new_l2_to_l1_msgs : BoundedVec<Field, MAX_NEW_L2_TO_L1_MSGS_PER_TX>,

    encrypted_logs_hash : [Field; NUM_FIELDS_PER_SHA256],
    unencrypted_logs_hash : [Field; NUM_FIELDS_PER_SHA256],

    // Here so that the gas cost of this request can be measured by circuits, without actually needing to feed in the
    // variable-length data.
    encrypted_log_preimages_length : Field,
    unencrypted_log_preimages_length : Field,

    new_contracts : BoundedVec<NewContractData, MAX_NEW_CONTRACTS_PER_TX>,
    optionally_revealed_data : BoundedVec<OptionallyRevealedData, MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX>
}

struct FinalAccumulatedData{

    aggregation_object : AggregationObject,

    new_commitments : [Field; MAX_NEW_COMMITMENTS_PER_TX],
    new_nullifiers : [Field; MAX_NEW_NULLIFIERS_PER_TX],
    nullified_commitments : [Field; MAX_NEW_NULLIFIERS_PER_TX],

    // For pending nullifiers, we have:
    // nullifiedCommitments[j] != 0 <==> newNullifiers[j] nullifies nullifiedCommitments[j]

    private_call_stack : [Field; MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX],
    public_call_stack : [Field; MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX],
    new_l2_to_l1_msgs : [Field; MAX_NEW_L2_TO_L1_MSGS_PER_TX],

    encrypted_logs_hash : [Field; NUM_FIELDS_PER_SHA256],
    unencrypted_logs_hash : [Field; NUM_FIELDS_PER_SHA256],

    // Here so that the gas cost of this request can be measured by circuits, without actually needing to feed in the
    // variable-length data.
    encrypted_log_preimages_length : Field,
    unencrypted_log_preimages_length : Field,

    new_contracts : [NewContractData; MAX_NEW_CONTRACTS_PER_TX],
    optionally_revealed_data : [OptionallyRevealedData; MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX]
}
