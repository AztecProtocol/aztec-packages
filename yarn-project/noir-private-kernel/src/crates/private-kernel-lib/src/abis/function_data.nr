use crate::abis::function_selector::FunctionSelector;

struct FunctionData {
    // First four bytes of the abi encoding 
    // of a function. 
    selector : FunctionSelector,
    is_internal : bool,
    is_private : bool,
    is_constructor : bool,
}


impl FunctionData {
    // A field is ~256 bits
    // TODO(hash): Since, function data can fit into a Field,
    // This method will simply return a bit packed Field instead of hashing
    fn hash(self) -> Field {
        0
    }
}
