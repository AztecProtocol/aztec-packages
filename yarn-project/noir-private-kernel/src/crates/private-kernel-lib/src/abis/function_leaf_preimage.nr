use crate::abis::function_selector::FunctionSelector;
use dep::aztec::constants_gen;

struct FunctionLeafPreimage {
    selector : FunctionSelector,
    is_internal : bool,
    is_private : bool,
    vk_hash : Field,
    acir_hash : Field
}

impl FunctionLeafPreimage {
    fn hash(self) -> Field {
        let _hash_index = constants_gen::GENERATOR_INDEX__FUNCTION_LEAF;
        0
    }
}
