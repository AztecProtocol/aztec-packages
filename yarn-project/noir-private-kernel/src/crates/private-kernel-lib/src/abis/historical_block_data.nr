use crate::block::Block;

struct HistoricalBlockData {
    blocks_tree_root : Field,
    block : Block,
    // Private data
    // This is marked in the cpp code as an enhancement
    private_kernel_vk_tree_root : Field,
}

impl HistoricalBlockData {
    fn assert_is_zero(self) {
        self.block.assert_is_zero();
        assert(self.private_kernel_vk_tree_root == 0);
    }

    // TODO: We can move this onto block and then 
    fn to_array(self) -> [Field;7] {
        // This comment was copied from the cpp codebase.
        //
        // TODO: Note private_kernel_vk_tree_root, is not included yet as
        // it is not present in noir,
        [
            self.block.private_data_tree_root,
            self.block.nullifier_tree_root,
            self.block.contract_tree_root,
            self.block.l1_to_l2_data_tree_root,
            self.blocks_tree_root,
            self.block.public_data_tree_root,
            self.block.global_variables_hash
        ]
    }

    // Note: private_kernel_vk_tree_root is not being hashed here
    // TODO(David): Is this because it is not included in Noir?
    fn hash(self) -> Field {
        // Also note that the blocks_tree_root does not get hashed
        self.block.hash()
    }
    
    fn private_data_tree_root(self) -> Field {
        self.block.private_data_tree_root
    }

    fn contract_tree_root(self) -> Field {
        self.block.contract_tree_root
    }
}
