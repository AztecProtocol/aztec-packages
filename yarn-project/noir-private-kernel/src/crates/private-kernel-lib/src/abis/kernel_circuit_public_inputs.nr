use crate::abis::combined_constant_data::CombinedConstantData;
use crate::abis::combined_accumulated_data::{
    CombinedAccumulatedData,
    FinalAccumulatedData,
    CombinedAccumulatedDataBuilder,
};

struct KernelCircuitPublicInputs {
    end : CombinedAccumulatedData,
    constants : CombinedConstantData,

    is_private : bool,
}

struct KernelCircuitPublicInputsFinal {
    end : FinalAccumulatedData,
    constants : CombinedConstantData,
    is_private : bool,
}

struct KernelCircuitPublicInputsBuilder {
    end : CombinedAccumulatedDataBuilder,
    constants : CombinedConstantData,

    is_private : bool,
}

impl KernelCircuitPublicInputsBuilder {
    pub fn finish(self) -> KernelCircuitPublicInputs {
        KernelCircuitPublicInputs {
            end : self.end.finish(),
            constants : self.constants,
            is_private : self.is_private,
        }
    }

    pub fn to_final(self) -> KernelCircuitPublicInputsFinal {
        KernelCircuitPublicInputsFinal {
            end : self.end.to_final(),
            constants : self.constants,
            is_private : self.is_private,
        }
    }
}
