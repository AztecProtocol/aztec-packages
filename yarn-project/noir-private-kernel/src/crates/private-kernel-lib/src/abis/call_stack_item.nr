use crate::address::Address;
use crate::abis::function_data::FunctionData;
use dep::aztec::constants_gen::{MAX_READ_REQUESTS_PER_CALL};
use crate::abis::private_circuit_public_inputs::PrivateCircuitPublicInputs;
use crate::abis::public_circuit_public_inputs::PublicCircuitPublicInputs;
use dep::aztec::constants_gen;

// TODO(Noir-bug): These should be type aliases

struct PublicCallStackItem {
    inner : CallStackItem<PublicCircuitPublicInputs>
}

struct PrivateCallStackItem {
    inner : CallStackItem<PrivateCircuitPublicInputs>
}

// Parametrize function over the public inputs.
// The public inputs can be from either a public or a private circuit
// TODO: We could monomorphize this struct and create a PrivateCallStackItem and 
// TODO: a PublicCallStackItem
struct CallStackItem<AppCircuitPublicInputs> {
    // Comment copied from the cpp codebase
    //
    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    contract_address : Address,

    public_inputs : AppCircuitPublicInputs,

    // True if this call stack item represents a request to execute a function rather than a
    // fulfilled execution. Used when enqueuing calls from private to public functions.
    is_execution_request : bool,

    function_data : FunctionData,
}

// TODO(Traits): Defining a hash trait bound
// Would allow the impl to be over `T : Hash`
// Where T is public inputs

// Noir(bug): If we can use a type alias instead of a new struct
// then we can remove the getter methods below for things like
// public_inputs and contract_address
impl PublicCallStackItem {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.inner.contract_address.to_field(),
            self.inner.function_data.hash(),
            self.inner.public_inputs.hash(),
        ], constants_gen::GENERATOR_INDEX__CALL_STACK_ITEM)[0]
    }

    fn public_inputs(self) -> PublicCircuitPublicInputs {
        self.inner.public_inputs
    }

    fn contract_address(self) -> Address{
        self.inner.contract_address
    }

    fn function_data(self) -> FunctionData{
        self.inner.function_data
    }
}
impl PrivateCallStackItem {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.inner.contract_address.to_field(),
            self.inner.function_data.hash(),
            self.inner.public_inputs.hash(),
        ], constants_gen::GENERATOR_INDEX__CALL_STACK_ITEM)[0]
    }

    fn public_inputs(self) -> PrivateCircuitPublicInputs {
        self.inner.public_inputs
    }

    fn contract_address(self) -> Address{
        self.inner.contract_address
    }

    fn function_data(self) -> FunctionData{
        self.inner.function_data
    }
}
