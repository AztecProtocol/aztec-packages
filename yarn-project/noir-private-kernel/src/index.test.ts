import {
  CircuitsWasm,
  ContractDeploymentData,
  EthAddress,
  FunctionData,
  FunctionLeafPreimage,
  FunctionSelector,
  Point,
  TxContext,
  TxRequest,
  makeAztecAddress,
  makePoint,
  makePrivateKernelInputsInit,
  makeSelector,
} from '@aztec/circuits.js';
import { computeCompleteAddress, computeFunctionLeaf, computeTxHash } from '@aztec/circuits.js/abis';
import { Fr } from '@aztec/foundation/fields';
import { DebugLogger, createDebugLogger } from '@aztec/foundation/log';

import { executeInit } from './index.js';

describe('Private kernel', () => {
  let logger: DebugLogger;
  beforeAll(() => {
    logger = createDebugLogger('noir-private-kernel');
  });

  it.skip('Executes private kernel init circuit with abi all zeroes (does not crash)', async () => {
    logger('Initialized Noir instance with private kernel init circuit');

    const kernelInputs = makePrivateKernelInputsInit();

    // TODO(Kev): How was this passing with the circuits.wasm version? Maybe it never tested init with mock?
    // Delegate and static calls are not allowed in the init
    kernelInputs.privateCall.callStackItem.publicInputs.callContext.isDelegateCall = false;
    kernelInputs.privateCall.callStackItem.publicInputs.callContext.isStaticCall = false;

    // Tx origin address must equal the storage contract address on the call stack
    kernelInputs.privateCall.callStackItem.contractAddress =
      kernelInputs.privateCall.callStackItem.publicInputs.callContext.storageContractAddress;
    kernelInputs.txRequest.origin = kernelInputs.privateCall.callStackItem.contractAddress;

    // Tx request function data and args hash must match
    kernelInputs.txRequest.functionData = kernelInputs.privateCall.callStackItem.functionData;
    kernelInputs.txRequest.argsHash = kernelInputs.privateCall.callStackItem.publicInputs.argsHash;

    // Make these 0 so that the circuit does not try to apply constraints on them
    kernelInputs.privateCall.callStackItem.publicInputs.privateCallStack.fill(Fr.ZERO);
    kernelInputs.privateCall.callStackItem.publicInputs.readRequests.fill(Fr.ZERO);

    const _kernelOutputs = await executeInit(kernelInputs);

    logger('Executed private kernel init circuit with all zeroes');
  });
});

describe('Noir compatibility tests (interop_testing.nr)', () => {
  // Tests in this file are to check that what we are computing in Noir
  // is equivalent to what we were computing in circuits.js/the typescript implementation
  // This is to ensure that we have not introduced any bugs in the transition from circuits.js to Noir
  let logger: DebugLogger;
  beforeAll(() => {
    logger = createDebugLogger('noir-private-kernel-compatibility');
  });

  it('Complete Address matches Noir', async () => {
    logger('Initialized Noir instance with private kernel init circuit');
    const wasm = await CircuitsWasm.get();
    const deployerPubKey = makePoint();
    const contractAddrSalt = new Fr(3n);
    const treeRoot = new Fr(4n);
    const constructorHash = new Fr(5n);

    const res = computeCompleteAddress(wasm, deployerPubKey, contractAddrSalt, treeRoot, constructorHash);

    expect(res.address.toString()).toMatchSnapshot();
    expect(res.publicKey).toMatchSnapshot();
    expect(res.partialAddress.toString()).toMatchSnapshot();
  });

  it('TxRequest Hash matches Noir', async () => {
    const wasm = await CircuitsWasm.get();

    const deploymentData = new ContractDeploymentData(
      new Point(new Fr(1), new Fr(2)),
      new Fr(1),
      new Fr(2),
      new Fr(3),
      new EthAddress(numberToBuffer(1)),
    );
    const txRequest = TxRequest.from({
      origin: makeAztecAddress(1),
      functionData: new FunctionData(makeSelector(2), false, true, true),
      argsHash: new Fr(3),
      txContext: new TxContext(false, false, true, deploymentData, Fr.ZERO, Fr.ZERO),
    });
    const hash = computeTxHash(wasm, txRequest);

    expect(hash.toString()).toMatchSnapshot();
  });

  it('ComputeContractAddressFromPartial matches Noir', async () => {
    const wasm = await CircuitsWasm.get();

    const deploymentData = new ContractDeploymentData(
      new Point(new Fr(1), new Fr(2)),
      new Fr(1),
      new Fr(2),
      new Fr(3),
      new EthAddress(numberToBuffer(1)),
    );
    const txRequest = TxRequest.from({
      origin: makeAztecAddress(1),
      functionData: new FunctionData(makeSelector(2), false, true, true),
      argsHash: new Fr(3),
      txContext: new TxContext(false, false, true, deploymentData, Fr.ZERO, Fr.ZERO),
    });
    const hash = computeTxHash(wasm, txRequest);

    expect(hash.toString()).toMatchSnapshot();
  });

  it('Function leaf matches noir', async () => {
    const wasm = await CircuitsWasm.get();

    const fnLeafPreimage = new FunctionLeafPreimage(new FunctionSelector(27), false, true, new Fr(1), new Fr(2));
    const fnLeaf = computeFunctionLeaf(wasm, fnLeafPreimage);
    expect(fnLeaf.toString()).toMatchSnapshot();
  });
});

function numberToBuffer(value: number) {
  // This can be used to convert a number to a buffer
  // and used as an EthAddress or AztecAddress.
  //
  // I think the EthAddress taking in 32 bytes is
  // not great, but I'll take advantage of it here.
  return new Fr(value).toBuffer();
}
