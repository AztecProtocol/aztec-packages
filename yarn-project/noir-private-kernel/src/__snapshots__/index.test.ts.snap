// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Noir compatibility tests (interop_testing.nr) Complete Address matches Noir 1`] = `"0x1a6e89b034478713c7a9f1c77fb80af995f708f6f208bf352b4dda2124739109"`;

exports[`Noir compatibility tests (interop_testing.nr) Complete Address matches Noir 2`] = `
Point {
  "kind": "point",
  "x": Fr {
    "value": 1n,
  },
  "y": Fr {
    "value": 2n,
  },
}
`;

exports[`Noir compatibility tests (interop_testing.nr) Complete Address matches Noir 3`] = `"0x197673f31940878b2d6c681223dbed9cfacd2f722cbe30155225b2ada17778db"`;

exports[`Noir compatibility tests (interop_testing.nr) ComputeContractAddressFromPartial matches Noir 1`] = `"0x0b487ff2900ae1178e131bfe333fdbc351beef658f7c0d62db2801429b1aab75"`;

exports[`Noir compatibility tests (interop_testing.nr) Function leaf matches noir 1`] = `"0x1ad8ece7f40e63d011ae47c6ce6cdaf31d632a23f5cf35bbeaaf69c8302afdbc"`;

exports[`Noir compatibility tests (interop_testing.nr) TxRequest Hash matches Noir 1`] = `"0x0b487ff2900ae1178e131bfe333fdbc351beef658f7c0d62db2801429b1aab75"`;
